okapi-core/src/main/java/org/folio/okapi/util/DepResolution.java

```diff
@@ -788,4 +796,132 @@ static void sortTenantModules(List<TenantModuleDescriptor> tml,
+              + permissionName + "' in permissionsRequired");
+        }
+      }
+    }
+    for (Map.Entry<String,Set<ModuleDescriptor>> ent : referSubPermissions.entrySet()) {
+      String permissionName = ent.getKey();
+      if (!defined.containsKey(permissionName) && !unknown.contains(permissionName)) {
+        for (ModuleDescriptor md : ent.getValue()) {
+          errors.computeIfAbsent(md, x -> new ArrayList<>()).add("Undefined permission '"
+              + permissionName + "' in subPermissions");
+        }
+      }
+    }
+    for (Map.Entry<String,Set<ModuleDescriptor>> ent: defined.entrySet()) {
+      if (ent.getValue().size() > 1) {
+        String permissionName = ent.getKey();
+        String names = ent.getValue()
+            .stream().map(x -> x.getId()).collect(Collectors.joining(", "));
+        for (ModuleDescriptor md : ent.getValue()) {
+          errors.computeIfAbsent(md, x -> new ArrayList<>()).add("Permission '"
+              + permissionName + "' defined in multiple modules: " + names);
+        }
+      }
+    }
+    return errors;
+  }
+
+  private static void getReferPermissions(
+      ModuleDescriptor md,
+      Map<String, Set<ModuleDescriptor>> referRequired,
+      Map<String, Set<ModuleDescriptor>> referModulePermissions,
+      Map<String, Set<ModuleDescriptor>> referSubPermissions,
+      Set<String> unknownPermissions) {
+
+    for (InterfaceDescriptor descriptor : md.getProvidesList()) {
+      RoutingEntry[] handlers = descriptor.getHandlers();
+      if (!descriptor.isRegularHandler() || handlers == null) {
+        continue;
+      }
+      for (RoutingEntry re : handlers) {
+        String[] modulePermissions = re.getModulePermissions();
+        if (modulePermissions != null) {
+          for (String modulePermission : modulePermissions) {
+            referModulePermissions.computeIfAbsent(modulePermission, k -> new HashSet<>()).add(md);
+            if (unknownPermissions != null) {
+              unknownPermissions.add(modulePermission);
+            }
+          }
+        }
+        for (String requiredPermission : re.getPermissionsRequired()) {
+          referRequired.computeIfAbsent(requiredPermission, k -> new HashSet<>()).add(md);
+          if (unknownPermissions != null) {
+            unknownPermissions.add(requiredPermission);
+          }
+        }
+      }
+    }
+    Permission[] permissionSets = md.getPermissionSets();
+    if (permissionSets != null) {
+      for (Permission permission: permissionSets) {
+        String[] subPermissions = permission.getSubPermissions();
+        if (subPermissions != null) {
+          for (String subPermission : subPermissions) {
+            referSubPermissions.computeIfAbsent(subPermission, k -> new HashSet<>()).add(md);
+            if (unknownPermissions != null) {
+              unknownPermissions.add(subPermission);
+            }
+          }
+        }
+      }
+    }
+  }
+
+  private static void getDefinedPermissions(
```

Comment:
`addPermissions` might be a better name because it returns void and doesn't get from `defined` but writes into it.

Reply:
ok